theseus/optimizer/sparse_linearization.py

```diff
@@ -174,3 +180,16 @@ def Av(self, v: torch.Tensor) -> torch.Tensor:
         return sparse_mv(
             self.num_cols, A_row_ptr, A_col_ind, self.A_val.double(), v.double()
         ).to(v.dtype)
+
+    def diagonal_scaling(self, v: torch.Tensor) -> torch.Tensor:
+        assert v.ndim == 2
+        assert v.shape[1] == self.num_cols
+        if self._AtA_diag is None:
```

Comment:
I'm wondering if this is a bug. In a batch, some instances might be updated while some remain the same. Therefore, do we need a mask for this?

Reply:
I don't think this should cause incorrect results/behavior, but for sure some duplicated computation is possible. And this is true at the linearization level, not just for this method; meaning linearization is run for the full batch every time linearization needs to be called. And, while the `Vectorize` caches jacobians/errors results, as soon as a single batch index is updated for *any* variable, vectorization is recomputed again, which is the largest part of the cost.  

But at least in terms of the backward pass running without issues, this should be OK, since we have been using this masking logic for a long time without spotting any issues. You can think of this as just recomputing the scaling with the tensors at the tip of the compute graph. 

That being said, I think it's interesting to add an issue for considering a more selective linearization logic, it might be worth adding at some point.